### PR TITLE
HDDS-11304. Make up for the missing functionality in CommandDispatcher

### PR DESCRIPTION
**Pull request by jianghuazhu**
## What changes were proposed in this pull request?
In CommandDispatcher, there are some codes that need to be cleaned up. In addition, there are some missing functions related to CommandDispatcher, and we should make up for them as much as possible.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-11304

## How was this patch tested?
It is necessary to ensure that CI passes normally.
After adding this PR, jmx will have some new changes.
Hadoop:service=HddsDatanode,name=CommandHandlerMetrics
Old:
image

New:
image

### CODE REVIEW

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java [hunk 1]
```diff
@@ -102,7 +98,7 @@ public void handle(SCMCommand command) {
         LOG.error("Exception while handle command, ", ex);
       }
     } else {
-      LOG.error("Unknown SCM Command queued. There is no handler for this " +
+      LOG.warn("Unknown SCM Command queued. There is no handler for this " +
```
Comment (kerneltime) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java:101]: Why is this not an `ERROR`

**Comment by adoroszlai** [comment]
Thanks @jianghuazhu for the patch.  Please update the title and description, since this does not seem to be a "cleanup".  Please also describe how the change was tested.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java (hunk 1 above)
Comment (jianghuazhu) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java:101]: Thanks @kerneltime  for the comment and review.
No exception is generated here, so I think setting the log level to `warn` is more appropriate here. What do you think?

**Comment by jianghuazhu** [comment]
> Thanks @jianghuazhu for the patch. Please update the title and description, since this does not seem to be a "cleanup". Please also describe how the change was tested.

Thanks @adoroszlai  for the comment and review.
I will update the title name later. Also, do you think additional tests need to be added?

**Comment by jianghuazhu** [comment]
@adoroszlai , I have updated the PR description and test description, and added some unit tests.
Can you review it again?
Thanks.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java (hunk 1 above)
Comment (kerneltime) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java:101]: SCM getting an unknown command implies that either something during upgrade failed or a service is misconfigured. I would keep it as `ERROR`

Comment (adoroszlai) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java:101]: +1 for keeping it as error.

Comment (jianghuazhu) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/commandhandler/CommandDispatcher.java]: Thanks @kerneltime  and @adoroszlai  for the explanation.
I have updated it.

**Comment by jianghuazhu** [comment]
@kerneltime , do you have any new suggestions?
Thanks.

**Review by fapifta — COMMENTED** [review]
Hi @jianghuazhu,
thank you for the contribution, overall it looks good.
I have a few questions/observations.


In DeleteBlockCommandHandler, we use ThreadPoolExecutor as the type of the executor, in DeleteContainerCommandHandler and in CloseContainerCommandHandler we use ExecutorService as the type. In all cases we create a ThreadPoolExecutor anyways.
Can you please change the type of the private executor field to ThreadPoolExecutor and avoid the case in the thread pool size related getters? 

This is more of a question, I don't know if it makes sense, but in the TestCloseContainerCommandHandler you have added a test that checks the thread pool's actual size and max size. This seems to be static pretty much and it is because the subject seems to be initialized to the CloseContainerHandler with a pool size of 1. If we want to better test this, we should check what happens if we change this configuration, and whether this is respected by the handlers properly... (This may be in some test related to DataNodeStateMachine, as that is the one instantiates these handlers. At the moment it seems to be working well, however tests for these configs would give us a warning if it is changed, and these can be tested now as the threadpool size and max size is exposed for the handlers.)

In the TestReplicationSupervisor, you add a new test, that asserts on the number of different tasks based on the class of the task. This can be done because the request and total time counter was updated to be a map, and to contain task specific metrics.
This is really good, but it does not seem to be complete... With this addition, we can have metrics for different tasks in the ReplicationSupervisor, but we do not have it at the end of the day...
I would suggest either add this separately, and extend this to all the metrics (success count, failure count, timout count, and skip count) besides the request count and total time.
Once that is done, we can add separate metrics for all task types in the ReplicationSupervisorMetrics class' getMetrics method, to expose these metrics in the JMX output of the DataNode.
If we decide to implement this as part of this PR, I am fine with that, please reflect that in the PR, if we are to add a new PR to cover this, then it would be better to leave this out here.

**Comment by jianghuazhu** [comment]
Thank you very much @fapifta  for your comments and review.
Your suggestions are very meaningful.
I will update it later and make two changes:
1. In DeleteContainerCommandHandler and CloseContainerCommandHandler, I will define the thread pool type as ThreadPoolExecutor.
2. I will create a new issue to improve ReplicationSupervisor.

**Review by fapifta — APPROVED** [review]
Thank you for addressing the comments @jianghuazhu, can you please link the new JIRA for the ReplicationSupervisor changes to HDDS-11304 and here in a comment, so if one wants to track the history it is easy to follow?

The change look good to me, +1, pending CI.

**Comment by fapifta** [comment]
It seems that the change from direct executor to ThreadPoolExecutor causes some trouble in the tests.

I would guess adding Thread.sleep would solve the problem, but may I suggest to use some wait mechanism instead of just sleeping for some time?
Maybe we can use a lock that is being locked before the handler.handle call, and being unlocked in the afterExecute method in the ThreadPoolExecutor (we need to sub-class it to specify afterExecute). This would better avoid flakiness then simply call thread sleep, and will also reduce the test runtime.

**Comment by jianghuazhu** [comment]
@fapifta , I have updated it.
ci: https://github.com/jianghuazhu/ozone/actions/runs/10598477539

HDDS-11376 is about improving ReplicationSupervisor.

**Review by fapifta — APPROVED** [review]
Thank you for your continued efforts on this @jianghuazhu, the change looks good, I restarted the failed CI job, the failure seems to be unrelated, but let's wait until it gets green.

**Pull request merged**